src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonDocument.MetadataDb.cs

```diff
@@ -88,6 +88,23 @@ private struct MetadataDb : IDisposable
             private const int SizeOrLengthOffset = 4;
             private const int NumberOfRowsOffset = 8;
 
+            private static readonly MetadataDb LockedTrue =
+                CreateLockedForNonStringPrimitiveImpl(JsonTokenType.True, JsonConstants.TrueValue.Length);
+
+            private static readonly MetadataDb LockedFalse =
+                CreateLockedForNonStringPrimitiveImpl(JsonTokenType.False, JsonConstants.FalseValue.Length);
+
+            private static readonly MetadataDb LockedNull =
+                CreateLockedForNonStringPrimitiveImpl(JsonTokenType.Null, JsonConstants.NullValue.Length);
+
+            // Index i is a singleton database for all numbers of length i
```

Comment:
What does "all numbers of length i" mean exactly? Is it that we're caching all possible numbers that are up to 8 characters long? That would be ~ $10^8$ different numbers, not accounting for decimal points or exponentials.

Reply:
No, the metadata database doesn't store the actual content, only the index into the actual content. The document will still need to store the UTF-8. But for primitive JSON values (like string and number), the metadata is just "what is the start offset of the value" and "how long is the value" (and "is the value escaped" for strings, but I chose not to cache escaped strings). The start offset is always 0 for number and 1 for string (to skip the quote) so we just need to store "how long is the value". So there would only be 8 cached `MetadataDb`s which represent 0 <= n <10^8 actual values.

Reply:
I see now. I take it these are only used if the strings or numbers don't have preceding or trailing whitespace or comments?

Reply:
Added tests.